druid/src/text/format.rs

```diff
@@ -0,0 +1,208 @@
+    /// regardless of success or failure, include new text and selection values
+    /// that should replace the current ones.
+    ///
+    ///
+    /// # Replacing the text or selection during validation
+    ///
+    /// Your `Formatter` may wish to change the current text or selection during
+    /// editing for a number of reasons. For instance if validation fails, you
+    /// may wish to allow editing to continue, but select the invalid region;
+    /// alternatively you may consider input valid but want to transform it,
+    /// such as by changing case or inserting spaces.
+    ///
+    /// If you do *not* explicitly set replacement text, and validation is not
+    /// successful, the edit will be ignored.
+    ///
+    /// [`Validation`]: Validation
+    fn validate_partial_input(&self, input: &str, sel: &Selection) -> Validation;
+
+    /// The value represented by the input, or an error if the input is invalid.
+    ///
+    /// This must return `Ok()` for any string created by [`format`].
+    ///
+    /// [`format`]: #tymethod.format
+    fn value(&self, input: &str) -> Result<T, ValidationError>;
+}
+
+/// The result of a [`Formatter`] attempting to validate some partial input.
+///
+/// [`Formatter`]: Formatter
+pub struct Validation {
+    result: Result<(), ValidationError>,
+    /// A manual selection override.
+    ///
+    /// This will be set as the new selection (regardless of whether or not
+    /// validation succeeded or failed)
+    pub selection_change: Option<Selection>,
+    /// A manual text override.
+    ///
+    /// This will be set as the new text, regardless of whether or not
+    /// validation failed.
+    pub text_change: Option<String>,
+}
+
+/// An error returned by a [`Formatter`] when it cannot parse input.
+///
+/// This implements [`source`] so you can access the inner error type.
+///
+/// [`source`]: std::Error::source;
+// This is currently just a wrapper around the inner error; in the future
+// it may grow to contain information such as invalid spans?
+// TODO: the fact that this uses `Arc` to work with `Data` is a bit inefficient;
+// it means that we will update when a new instance of an identical error occurs.
+#[derive(Debug, Clone, Data)]
+pub struct ValidationError {
+    inner: Arc<dyn std::error::Error>,
+}
+
+/// A naive [`Formatter`] for types that implement [`FromStr`].
+///
+/// [`Formatter`]: Formatter
+/// [`FromStr`]: std::str::FromStr
+#[non_exhaustive]
+//TODO: take a fmt string (like `{:.2}` to allow user to customize output
+pub struct ParseFormatter;
+
+impl Validation {
+    /// Create a `Validation` indicating succes.
+    pub fn success() -> Self {
+        Validation {
+            result: Ok(()),
+            selection_change: None,
+            text_change: None,
+        }
+    }
+
+    /// Create a `Validation` with an error indicating the failure reason.
+    pub fn failure(err: impl std::error::Error + 'static) -> Self {
```

Comment:
Would also be nice if it had just a little documentation on what this does. Does this result in the edit being undone? Does something happen to the selection? Does the user see an error indicating they typed something wrong? (probably not the later 2 but it should be clear)

Reply:
This is a function that creates a `Validation` object with a given inner error. In the context of text editing, how the textbox reacts when validation fails is dependent on how that text box is configured.